File: src/com/prey/events/factories/EventFactory.java
package com.prey.events.factories;

import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;

import org.json.JSONObject;

import android.content.Context;
import android.content.Intent;
import android.net.ConnectivityManager;
import android.net.wifi.WifiManager;
//import android.os.BatteryManager;
import android.os.Bundle;
import android.provider.Settings;

import com.prey.PreyConfig;
import com.prey.PreyLogger;
import com.prey.beta.actions.PreyBetaController;
import com.prey.events.Event;
import com.prey.managers.PreyConnectivityManager;
import com.prey.managers.PreyTelephonyManager;

public class EventFactory {

	private static final String BOOT_COMPLETED="android.intent.action.BOOT_COMPLETED";
	private static final String CONNECTIVITY_CHANGE="android.net.conn.CONNECTIVITY_CHANGE";
	private static final String WIFI_STATE_CHANGED="android.net.wifi.WIFI_STATE_CHANGED";
//	private static final String ACTION_POWER_CONNECTED="android.intent.action.ACTION_POWER_CONNECTED";
//	private static final String ACTION_POWER_DISCONNECTED="android.intent.action.ACTION_POWER_DISCONNECTED";
//	private static final String ACTION_BATTERY_CHANGE="android.intent.action.ACTION_BATTERY_CHANGE";
//	private static final String ACTION_BATTERY_LOW="android.intent.action.ACTION_BATTERY_LOW";
	private static final String ACTION_SHUTDOWN="android.intent.action.ACTION_SHUTDOWN";
//	private static final String BATTERY_LOW="android.intent.action.BATTERY_LOW";
//	private static final String BATTERY_OKAY="android.intent.action.BATTERY_OKAY";
//	private static final String BATTERY_CHANGED="android.intent.action.BATTERY_CHANGED";
	
	private static final String AIRPLANE_MODE="android.intent.action.AIRPLANE_MODE";
	
//	private static final String USER_PRESENT="android.intent.action.USER_PRESENT";
	
	public static Event getEvent(Context ctx,Intent intent){
		String message="getEvent["+intent.getAction()+"]";
		//Toast.makeText(ctx, message, Toast.LENGTH_LONG).show();
		PreyLogger.d(message);
		if (BOOT_COMPLETED.equals(intent.getAction()) ){
			if (PreyConfig.getPreyConfig(ctx).isSimChanged()){
				JSONObject info=new JSONObject();
				try{
					info.put("new_phone_number",PreyTelephonyManager.getInstance(ctx).getLine1Number());
				}catch (Exception e) {
				}
				return new Event(Event.SIM_CHANGED,info.toString());
			}else{
				return new Event(Event.TURNED_ON);
			}
		}
		if (ACTION_SHUTDOWN.equals(intent.getAction()) ){
			return new Event(Event.TURNED_OFF);
		}
		if (CONNECTIVITY_CHANGE.equals(intent.getAction())  ){
			JSONObject info=new JSONObject();
			
			int wifiState = intent.getIntExtra(WifiManager.EXTRA_WIFI_STATE, WifiManager.WIFI_STATE_UNKNOWN);
			PreyLogger.d("__wifiState:"+wifiState);;
			
			if (!PreyConnectivityManager.getInstance(ctx).isWifiConnected()) {
			  Bundle extras = intent.getExtras();
		        if (extras!=null){
		               if("connected".equals(extras.getString(ConnectivityManager.EXTRA_REASON))){
		            	   try{Thread.sleep(2000);}catch(Exception e){}
		            	   PreyConfig.getPreyConfig(ctx).registerC2dm();
		               }
		        }
			}
			try{
				if (!PreyConnectivityManager.getInstance(ctx).isMobileConnected()) {
					info.put("connected", "mobile");
					if (wifiState == WifiManager.WIFI_STATE_ENABLED) {
						try{Thread.sleep(2000);}catch(Exception e){}
						PreyConfig.getPreyConfig(ctx).registerC2dm();
					}
				}
			}catch (Exception e) {
			}
			return new Event(Event.WIFI_CHANGED,info.toString());
		}
		if (  WIFI_STATE_CHANGED.equals(intent.getAction()) ){
			JSONObject info=new JSONObject();
			int wifiState = intent.getIntExtra(WifiManager.EXTRA_WIFI_STATE, WifiManager.WIFI_STATE_UNKNOWN);
			PreyLogger.d("___wifiState:"+wifiState);;
			try{
				if (wifiState == WifiManager.WIFI_STATE_ENABLED) {
					info.put("connected", "wifi");
					try{Thread.sleep(2000);}catch(Exception e){}
					PreyConfig.getPreyConfig(ctx).registerC2dm();
				}
			}catch (Exception e) {
			}
			return new Event(Event.WIFI_CHANGED,info.toString());
		}
		if (AIRPLANE_MODE.equals(intent.getAction())){
			if(!isAirplaneModeOn(ctx)){
				try{Thread.sleep(4000);}catch(Exception e){}
				PreyBetaController.startPrey(ctx);
				PreyConfig.getPreyConfig(ctx).registerC2dm();
			}
		}
		/*
		if (ACTION_POWER_CONNECTED.equals(intent.getAction()) ){
			return new Event(Event.PLUGGED);
		}
		if (ACTION_POWER_DISCONNECTED.equals(intent.getAction()) ){
			return new Event(Event.UN_PLUGGED);
		}
			
		if (ACTION_BATTERY_LOW.equals(intent.getAction()) && isValidLowBattery(ctx)){
			return new Event(Event.BATTERY_LOW);
		}
		if (BATTERY_LOW.equals(intent.getAction()) && isValidLowBattery(ctx)){
			return new Event(Event.BATTERY_LOW);
		}
		if (BATTERY_OKAY.equals(intent.getAction()) ){
			return new Event(Event.BATTERY_OK);
		}
		if (BATTERY_CHANGED.equals(intent.getAction()) ){
			return new Event(Event.BATTERY_CHANGE);
		}
		if (ACTION_BATTERY_CHANGE.equals(intent.getAction())){
			int plugged = intent.getIntExtra(BatteryManager.EXTRA_PLUGGED, -1);
			JSONObject info=new JSONObject();
			try{
				if(plugged==BatteryManager.BATTERY_PLUGGED_AC ){
					info.put("source ","ac");
				}else{
					if(plugged==BatteryManager.BATTERY_PLUGGED_USB){
						info.put("source ","usb");
					}
				}
			}catch (Exception e) {
			}
			//return new Event(Event.PLUGGED,info.toString());
		}*/
		return null;
	}
	
	public static boolean isAirplaneModeOn(Context context){
		   return Settings.System.getInt(context.getContentResolver
		                        (),Settings.System.AIRPLANE_MODE_ON, 0) != 0;
	}
	
	private static SimpleDateFormat sdf=new SimpleDateFormat("dd/MM/yy hh:mm:ss",Locale.getDefault());
	
	public static boolean isValidLowBattery(Context ctx) {
		try{
			Calendar cal=Calendar.getInstance();
			cal.setTime(new Date());
			cal.add(Calendar.HOUR,-3);
			long leastThreeHours=cal.getTimeInMillis();
			long lowBatteryDate=PreyConfig.getPreyConfig(ctx).getLowBatteryDate();
			PreyLogger.d("lowBatteryDate :"+lowBatteryDate+" "+sdf.format(new Date(lowBatteryDate)));
			PreyLogger.d("leastMinutes   :"+leastThreeHours+" "+sdf.format(new Date(leastThreeHours)));
			if(lowBatteryDate==0||leastThreeHours>lowBatteryDate){
				long now=new Date().getTime();
				PreyConfig.getPreyConfig(ctx).setLowBatteryDate(now);
				return true;
			}
			return false;
		}catch(Exception e){
			return false;
		}
	}
}

File: app/src/main/java/com/prey/events/factories/kotlin/EventFactory.kt
/*******************************************************************************
 * Created by Orlando Aliaga
 * Copyright 2025 Prey Inc. All rights reserved.
 * License: GPLv3
 * Full license at "/LICENSE"
 ******************************************************************************/
package com.prey.events.factories.kotlin

import android.content.Context
import android.content.Intent
import android.net.ConnectivityManager
import android.net.wifi.WifiManager
import android.provider.Settings
import com.prey.actions.aware.kotlin.AwareController
import com.prey.actions.location.kotlin.LocationUtil
import com.prey.actions.triggers.kotlin.BatteryTriggerReceiver
import com.prey.actions.triggers.kotlin.SimTriggerReceiver
import com.prey.beta.actions.kotlin.PreyBetaController
import com.prey.events.kotlin.Event
import com.prey.kotlin.PreyConfig
import com.prey.kotlin.PreyLogger
import com.prey.kotlin.PreyPermission
import com.prey.managers.kotlin.PreyConnectivityManager
import com.prey.net.kotlin.UtilConnection
import org.json.JSONObject
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale

object EventFactory {
    const val BOOT_COMPLETED: String = "android.intent.action.BOOT_COMPLETED"
    const val CONNECTIVITY_CHANGE: String = "android.net.conn.CONNECTIVITY_CHANGE"
    const val WIFI_STATE_CHANGED: String = "android.net.wifi.WIFI_STATE_CHANGED"
    const val ACTION_SHUTDOWN: String = "android.intent.action.ACTION_SHUTDOWN"
    const val AIRPLANE_MODE: String = "android.intent.action.AIRPLANE_MODE"
    const val BATTERY_LOW: String = "android.intent.action.BATTERY_LOW"
    const val SIM_STATE_CHANGED: String = "android.intent.action.SIM_STATE_CHANGED"
    const val USER_PRESENT: String = "android.intent.action.USER_PRESENT"
    const val ACTION_POWER_CONNECTED: String = "android.intent.action.ACTION_POWER_CONNECTED"
    const val ACTION_POWER_DISCONNECTED: String = "android.intent.action.ACTION_POWER_DISCONNECTED"
    const val LOCATION_MODE_CHANGED: String = "android.location.MODE_CHANGED"
    const val LOCATION_PROVIDERS_CHANGED: String = "android.location.PROVIDERS_CHANGED"
    const val NOTIFICATION_ID: Int = 888

    fun getEvent(ctx: Context, intent: Intent): Event? {
        val message = "getEvent[" + intent.action + "]"
        PreyLogger.d(message)
        if (BOOT_COMPLETED == intent.action) {
            notification(ctx)
            return Event(Event.TURNED_ON)
        }
        if (SIM_STATE_CHANGED == intent.action) {
            val state = intent.extras!!.getString(SimTriggerReceiver.EXTRA_SIM_STATE)
            if ("ABSENT" == state) {
                val info = JSONObject()
                try {
                    val simSerial = PreyConfig.getInstance(ctx).getSimSerialNumber()
                    if (simSerial != null && "" != simSerial) {
                        info.put("sim_serial_number", simSerial)
                    }
                } catch (e: Exception) {
                    PreyLogger.e("Error:" + e.message, e)
                }
                SimTriggerReceiver().onReceive(ctx, intent)
                return if (UtilConnection.getInstance().isInternetAvailable()) {
                    Event(Event.SIM_CHANGED, info.toString())
                } else {
                    null
                }
            }
        }
        if (LOCATION_PROVIDERS_CHANGED == intent.action || LOCATION_MODE_CHANGED == intent.action
        ) {
            object : Thread() {
                override fun run() {
                    sendLocationAware(ctx)
                }
            }.start()
        }
        if (ACTION_SHUTDOWN == intent.action) {
            return Event(Event.TURNED_OFF)
        }
        if (BATTERY_LOW == intent.action) {
            BatteryTriggerReceiver().onReceive(ctx, intent)
            return Event(Event.BATTERY_LOW)
        }
        if (ACTION_POWER_CONNECTED == intent.action) {
            BatteryTriggerReceiver().onReceive(ctx, intent)
            return Event(Event.POWER_CONNECTED)
        }
        if (ACTION_POWER_DISCONNECTED == intent.action) {
            BatteryTriggerReceiver().onReceive(ctx, intent)
            return Event(Event.POWER_DISCONNECTED)
        }
        if (CONNECTIVITY_CHANGE == intent.action) {
            return null
        }
        if (WIFI_STATE_CHANGED == intent.action) {
            val info = JSONObject()
            val wifiState =
                intent.getIntExtra(WifiManager.EXTRA_WIFI_STATE, WifiManager.WIFI_STATE_UNKNOWN)
            PreyLogger.d("getEvent ___wifiState:$wifiState")
            try {
                if (wifiState == WifiManager.WIFI_STATE_ENABLED) {
                    PreyLogger.d("getEvent wifiState connected")
                    info.put("connected", "wifi")
                    PreyBetaController.getInstance().startPrey(ctx)
                }
                if (wifiState == WifiManager.WIFI_STATE_DISABLED) {
                    PreyLogger.d("getEvent mobile connected")
                    info.put("connected", "mobile")
                }
            } catch (e: Exception) {
                PreyLogger.e("Error getEvent:" + e.message, e)
            }
            return Event(Event.WIFI_CHANGED, info.toString())
        }
        if (AIRPLANE_MODE == intent.action) {
            if (!isAirplaneModeOn(ctx)) {
                val verifyNotification = verifyNotification(ctx)
                if (!verifyNotification) {
                    notification(ctx)
                }
                var connected = false
                if (!PreyConnectivityManager.getInstance().isWifiConnected(ctx)) {
                    val extras = intent.extras
                    if (extras != null) {
                        if ("connected" == extras.getString(ConnectivityManager.EXTRA_REASON)) {
                            connected = true
                        }
                    }
                }
                if (!PreyConnectivityManager.getInstance().isMobileConnected(ctx)) {
                    val wifiState = intent.getIntExtra(
                        WifiManager.EXTRA_WIFI_STATE,
                        WifiManager.WIFI_STATE_UNKNOWN
                    )
                    if (wifiState == WifiManager.WIFI_STATE_ENABLED) {
                        connected = true
                    }
                }
                if (connected) {
                    PreyBetaController.getInstance().startPrey(ctx)
                }
            }
        }
        if (USER_PRESENT == intent.action) {
            PreyLogger.d("EventFactory USER_PRESENT")
            val minuteScheduled = PreyConfig.getInstance(ctx).getMinuteScheduled()
            if (minuteScheduled > 0) {
                PreyBetaController.getInstance().startPrey(ctx, null)
            }
            return null
        }
        return null
    }

    fun sendLocationAware(ctx: Context) {
        try {
            val isTimeLocationAware = PreyConfig.getInstance(ctx).isTimeLocationAware()
            PreyLogger.d("sendLocation isTimeLocationAware:$isTimeLocationAware")
            if (!isTimeLocationAware) {
                val locationNow = LocationUtil.getLocation(ctx, null, false)
                AwareController.getInstance().sendAware(ctx, locationNow)

                PreyConfig.getInstance(ctx).setTimeLocationAware()
            }
        } catch (e: Exception) {
            PreyLogger.e("Error sendLocation:" + e.message, e)
        }
    }

    fun isAirplaneModeOn(context: Context): Boolean {
        return Settings.System.getInt(
            context.contentResolver,
            Settings.System.AIRPLANE_MODE_ON,
            0
        ) != 0
    }

    private val sdf = SimpleDateFormat("dd/MM/yy hh:mm:ss", Locale.getDefault())

    fun isValidLowBattery(ctx: Context): Boolean {
        try {
            val cal = Calendar.getInstance()
            cal.time = Date()
            cal.add(Calendar.MINUTE, -1)
            val leastThreeHours = cal.timeInMillis
            val lowBatteryDate = PreyConfig.getInstance(ctx).getLowBatteryDate()
            PreyLogger.d("lowBatteryDate :" + lowBatteryDate + " " + sdf.format(Date(lowBatteryDate)))
            PreyLogger.d(
                "leastMinutes   :" + leastThreeHours + " " + sdf.format(
                    Date(
                        leastThreeHours
                    )
                )
            )
            if (lowBatteryDate == 0L || leastThreeHours > lowBatteryDate) {
                val now = Date().time
                PreyConfig.getInstance(ctx).setLowBatteryDate (now)
                return true
            }
            return false
        } catch (e: Exception) {
            return false
        }
    }

    /**
     * Method that returns if it has all the permissions
     *
     * @param ctx context
     * @return if you have all permissions
     */
    fun verifyNotification(ctx: Context?): Boolean {
        val canAccessCoarseLocation = PreyPermission.canAccessCoarseLocation(ctx)
        val canAccessFineLocation = PreyPermission.canAccessFineLocation(ctx)
        val canAccessStorage = PreyPermission.canAccessStorage(ctx)
        return (canAccessCoarseLocation || canAccessFineLocation) && canAccessStorage
    }

    /**
     * Method that opens the notification missing permissions
     *
     * @param ctx context
     */
    fun notification(ctx: Context?) {
    }
}